### PR TITLE
Radius: Implement Registration

### LINKED ISSUE #17: Implement /register

### PR DESCRIPTION
**Pull request by charliemirabile**
fixes #17

### CODE REVIEW

File: radius.py [hunk 1]
```diff
@@ -263,15 +263,16 @@ def expiry(self):
...
+    def body_args_query(self, key):
+        return html.escape(
+            decode(self.body_args.get(encode(key), [b''])[0]))
+
     # Attempt login using urelencoded credentials from request body
     def launch(self):
         new_ses = None
         if self.method == "POST":
-            def urldecode(key):
-                return html.escape(decode(self.body_args.get(encode(key),
```
Comment (theyoyojo) [inline radius.py:271]: nice good change

File: radius.py [hunk 2]
```diff
@@ -475,6 +475,7 @@ def handle_dashboard(rocket):
 def handle_stub(rocket, more=[]):
     meth_path = f'{rocket.method} {rocket.path_info}'
     content = f'<h3>Development stub for {meth_path} </h3>{"".join(more)}'
+    rocket.msg('oops')
```
Comment (theyoyojo) [inline radius.py:465]: Good point, silence is only appropriate in production

File: radius.py [hunk 3]
```diff
@@ -400,15 +400,6 @@ def mk_form_welcome(session):
...
 
 
-form_register = """
-    <form id="register" method="post" action="/register">
-        <label for="student_id">Student ID:</label>
-        <input name="student_id" type="text" id="student_id" /><br />
-        <button type="submit">Submit</button>
-    </form>
-""".strip()
-
```
Comment (theyoyojo) [inline radius.py:409]: valid

File: radius.py [hunk 4]
```diff
@@ -480,7 +480,12 @@ def handle_dashboard(rocket):
 
 
 def handle_register(rocket):
-    return handle_stub(rocket, [f'<code><br />{_OLD_NOTES}</code><br />'])
+    response_document = form_register
+    response_status = HTTPStatus.OK
+    rocket.msg('welcome, please register')
+    if rocket.method == 'POST':
+        return handle_stub(rocket, ['register in development, check back later'])  # NOQA: E501
+    return rocket.respond(response_status, response_document)
```
Comment (theyoyojo) [inline radius.py:508]: Confirmed that the page works as expected when HEAD is at this commit

File: init-db.sql [hunk 5]
```diff
@@ -5,8 +5,7 @@ CREATE TABLE users (
 	username string UNIQUE NOT NULL,
 	pwdhash string NOT NULL,
 	lfx boolean NOT NULL,
-	student_id integer
-);
```
Comment (theyoyojo) [inline init-db.sql:9]: no opposition from me

File: init-db.sql [hunk 6]
```diff
@@ -21,6 +21,11 @@ CREATE TABLE submissions (
 CREATE TABLE assignments (
 	web_id string PRIMARY KEY,
 	email_id string NOT NULL);
+CREATE TABLE newusers (
+	registration_id integer primary key,
+	student_id string UNIQUE NOT NULL,
+	username string UNIQUE NOT NULL,
```
Comment (theyoyojo) [inline init-db.sql:27]: absolutely registering right now

File: db.py [hunk 7]
```diff
@@ -247,7 +247,7 @@ def reg_getby_stuid(sid): return _set(REG_GETBY_STUID, sid)
 
 
 REG_DELBY_REGID = """
-DELETE FROM accounts
-WHERE id = ?;
+DELETE FROM newusers
+WHERE registration_id = ?;
```
Comment (theyoyojo) [inline db.py:251]: makes sense, this codepath was never tested before since it was never used

File: db.py [hunk 8]
```diff
@@ -232,8 +232,8 @@ def asn_get(): return _get(ASN_GET)
 # registration table inferface
 
 REG_INS = """
-INSERT VALUES username, password, student_id = (?,?,?)
-INTO accounts;
+INSERT INTO newusers (username, password, student_id)
+VALUES (?,?,?);
```
Comment (theyoyojo) [inline db.py:236]: see comment on previous commit

Comment (theyoyojo) [reply db.py:236]: this code was never run until today

File: hyperspace.py [hunk 9]
```diff
@@ -139,7 +139,10 @@ def do_newuser(args):
     if db.usr_getby_username(args.username)[0]:
         errx(f'cannot create duplicate user "{args.username}"')
     else:
-        db.usr_ins((args.username, do_bcrypt_hash(args, get=True), 0, 0))
+        db.usr_ins((args.username, do_bcrypt_hash(args, get=True),
+                    0, args.studentid or 0))
```
Comment (theyoyojo) [inline hyperspace.py:143]: interesting usage of or here but it works

File: db.py [hunk 10]
```diff
@@ -243,11 +243,11 @@ def reg_ins(tpl): return _set(REG_INS, tpl)
 FROM newusers
 WHERE student_id = ?;
 """.strip()
-def reg_getby_stuid(sid): return _set(REG_GETBY_STUID, sid)
+def reg_getby_stuid(sid): return _get(REG_GETBY_STUID, sid)
```
Comment (theyoyojo) [inline db.py:246]: sane

File: db.py [hunk 11]
```diff
@@ -243,11 +243,11 @@ def reg_ins(tpl): return _set(REG_INS, tpl)
...
 
 
 REG_DELBY_REGID = """
 DELETE FROM newusers
 WHERE registration_id = ?;
 """.strip()
-def reg_delby_regid(rid): return _get(REG_DELBY_REGID, rid)
+def reg_delby_regid(rid): return _set(REG_DELBY_REGID, rid)
```
Comment (theyoyojo) [inline db.py:253]: sane

File: radius.py [hunk 12]
```diff
@@ -479,12 +479,31 @@ def handle_dashboard(rocket):
 """.strip()
 
 
+register_response = '''
```
Comment (theyoyojo) [inline radius.py:482]: tiny nit: s/'''/"""

if you'd prefer, an alternative is to do treewide s/"""/'''

**Review by theyoyojo — CHANGES_REQUESTED** [review]
only change is requested in last commit

File: hyperspace.py (hunk 9 above)
Comment (charliemirabile) [reply hyperspace.py:143]: yeah, js behaves similarly, AND returns the first falsey value and OR returns the first truthy value

**Comment by theyoyojo** [comment]
Ran into a bug when registering:

```
spawned uWSGI http 1 (pid: 25)
Traceback (most recent call last):
  File "radius.py", line 550, in application
    return handle_register(rocket)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "radius.py", line 498, in handle_register
    response_document = register_response % {
                        ^^^^^^^^^^^^^^^^^^^^^
ValueError: unsupported format character '<' (0x3c) at index 104
[pid: 24|app: 0|req: 8/8] 127.0.0.1 () {62 vars in 1198 bytes} [Wed Jan 17 20:36:01 2024] POST /register => generated 0 bytes in 14 msecs (HTTP/1.0 500) 0 headers in 0 bytes (0 switches on core 0)

```

File: docker/docker-compose.yml [hunk 13]
```diff
@@ -5,9 +5,9 @@ services:
...
     ports:
-      - 80:80
-      - 443:443
-      - 465:465
-      - 995:995
+      - 1080:80
+      - 1443:443
+      - 1465:465
+      - 1995:995
       - 8448:8448
```
Comment (theyoyojo) [inline docker/docker-compose.yml:12]: drop this hunk, I assume this was committed accidentally

**Review by theyoyojo — CHANGES_REQUESTED** [review]

File: docker/docker-compose.yml (hunk 13 above)
Comment (charliemirabile) [reply docker/docker-compose.yml]: yup, I git added what I wanted, but then did `git commit -a --amend` anyways :facepalm:

Comment (theyoyojo) [reply docker/docker-compose.yml]: lgtm

**Review by theyoyojo — APPROVED** [review]
lgtm

**Pull request merged**